sql/common.go

```diff
@@ -0,0 +1,135 @@
+// Licensed to SkyAPM org under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. SkyAPM org licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sql
+
+import (
+	"context"
+	"database/sql/driver"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/SkyAPM/go2sky"
+	agentv3 "skywalking.apache.org/repo/goapi/collect/language/agent/v3"
+)
+
+const (
+	componentIDUnknown = 0
+	componentIDMysql   = 5010
+	componentIDSqlite  = 5011
+)
+
+const (
+	tagDbType          = "db.type"
+	tagDbInstance      = "db.instance"
+	tagDbStatement     = "db.statement"
+	tagDbSqlParameters = "db.sql.parameters"
+)
+
+type attribute struct {
+	dbType      DBType
+	peer        string
+	componentID int32
+}
+
+var ErrUnsupportedOp = errors.New("operation unsupported by the underlying driver")
+
+// emptyInjectFunc defines a empty injector for propagation.Injector function
+func emptyInjectFunc(key, value string) error { return nil }
+
+func newAttribute(name string, dbType DBType) attribute {
+	a := attribute{
+		dbType: dbType,
+	}
+	a.setPeer(name)
+	a.setComponentID()
+	return a
+}
+
+// setPeer parse dsn to a endpoint addr string (host:port)
+func (a attribute) setPeer(dsn string) {
+	var addr string
+	switch a.dbType {
+	case MYSQL:
+		// [user[:password]@][net[(addr)]]/dbname[?param1=value1&paramN=valueN]
+		re := regexp.MustCompile(`\(.+\)`)
+		addr = re.FindString(dsn)
+		addr = addr[1 : len(addr)-1]
+	case IPV4:
+		// ipv4 addr
+		re := regexp.MustCompile(`((2(5[0-5]|[0-4]\d))|[0-1]?\d{1,2})(\.((2(5[0-5]|[0-4]\d))|[0-1]?\d{1,2})){3}:\d{1,5}`)
+		addr = re.FindString(dsn)
+	}
+	a.peer = addr
+	return
+}
```

Comment:
I think it is not reasonable for the plugin to parse the dsn to get the address of the database, even for the same database of different plugins, there may also be different dsn format, perhaps the user should directly configure the target address?

Reply:
The default is to parse the dsn, it also supports configuration via `AttributeOption` func, [like this](https://github.com/SkyAPM/go2sky/blob/master/trace.go#L47). What do you think :)